File: ts/magicseam/stream.test.ts
// Copyright (C) 2025-2026 Malformed C. All rights reserved.
// SPDX-License-Identifier: BUSL-1.1

import { describe, expect, test } from "bun:test"
import { CAP_STREAM, OP_CALL, OP_STREAM, capsOffered, parseCaps, streamsAgreed } from "./stream"

// These live in their own file, separate from quic.test.ts, deliberately:
// quic.ts pulls in @matrixai/quic, which is not installed in every environment
// (it is not here), so a negotiation test placed there would silently not run.
// The negotiation logic has no transport dependency and should not inherit one.
describe("bulk-seam capability negotiation", () => {
  // The compatibility property the whole mechanism exists for. Guessing "yes"
  // would make this provider expect an opcode the consumer never writes,
  // misframing every call on the connection.
  test("an absent or empty caps frame means classic-only", () => {
    expect(streamsAgreed(parseCaps(undefined))).toBe(false)
    expect(streamsAgreed(parseCaps(new Uint8Array(0)))).toBe(false)
    expect(streamsAgreed([])).toBe(false)
  })

  test("a peer advertising the token negotiates it", () => {
    expect(streamsAgreed(parseCaps(Buffer.from(CAP_STREAM)))).toBe(true)
  })

  // Adding a capability later must not be a breaking change - which it would be
  // if an unrecognised token made a peer fail or mask a known one.
  test("unknown tokens do not mask a known one", () => {
    expect(streamsAgreed(parseCaps(Buffer.from("stream,quantum-teleport,zip")))).toBe(true)
    expect(streamsAgreed(parseCaps(Buffer.from("quantum-teleport")))).toBe(false)
  })

  test("stray separators and whitespace are tolerated", () => {
    expect(parseCaps(Buffer.from(" stream , , "))).toEqual([CAP_STREAM])
  })

  // This SDK must advertise exactly what it can serve: advertising nothing
  // leaves every TS provider classic-only; advertising more misframes calls.
  test("this SDK advertises the bulk seam", () => {
    expect(streamsAgreed(parseCaps(Buffer.from(capsOffered())))).toBe(true)
  })

  test("the opcodes are distinct", () => {
    expect(OP_CALL).not.toBe(OP_STREAM)
  })
})

File: go/magicseam/stream_test.go
// Copyright (C) 2025-2026 Malformed C. All rights reserved.
// SPDX-License-Identifier: BUSL-1.1

package magicseam

import (
	"testing"
)

// The compatibility property this whole mechanism exists for: a consumer that
// sends no capability frame must land on "classic only", never on "assume yes".
// Guessing yes would make this provider expect an opcode the consumer never
// writes, misframing every call on the connection.
func TestAbsentOrEmptyCapsMeansNoStreaming(t *testing.T) {
	if streamsAgreed(nil) {
		t.Error("no caps frame at all must not negotiate streaming")
	}
	if streamsAgreed(parseCaps(nil)) {
		t.Error("an absent frame must not negotiate streaming")
	}
	if streamsAgreed(parseCaps([]byte(""))) {
		t.Error("an empty frame must not negotiate streaming")
	}
}

func TestStreamingNegotiatesWhenAdvertised(t *testing.T) {
	if !streamsAgreed(parseCaps([]byte(CapStream))) {
		t.Error("a peer advertising stream must negotiate it")
	}
}

// A newer peer advertising tokens this build has never heard of must still
// negotiate the ones it does - otherwise adding a capability later becomes a
// breaking change, which is exactly what negotiation is here to prevent.
func TestUnknownTokensDoNotBreakAKnownOne(t *testing.T) {
	if !streamsAgreed(parseCaps([]byte("stream,quantum-teleport,zip"))) {
		t.Error("unknown tokens must not mask a known one")
	}
	if streamsAgreed(parseCaps([]byte("quantum-teleport"))) {
		t.Error("unknown tokens alone must not negotiate streaming")
	}
}

func TestCapsParsingTolerateStraySeparators(t *testing.T) {
	caps := parseCaps([]byte(" stream , , "))
	if len(caps) != 1 || caps[0] != CapStream {
		t.Errorf("parseCaps = %v, want exactly [%s] - empty tokens must be dropped", caps, CapStream)
	}
}

// This SDK must advertise what it can actually serve. Advertising nothing
// would silently leave every Go provider classic-only; advertising something
// it cannot serve would misframe calls.
func TestThisSDKAdvertisesTheBulkSeam(t *testing.T) {
	if !streamsAgreed(parseCaps([]byte(capsOffered(true)))) {
		t.Errorf("capsOffered(true) = %q, which does not advertise %q", capsOffered(true), CapStream)
	}
}

// The opcodes must differ, or every streaming call routes into the classic
// handler and vice versa.
func TestOpcodesAreDistinct(t *testing.T) {
	if opCall == opStream {
		t.Fatal("opCall and opStream must differ")
	}
}
